### PR TITLE
Add a "fast forward" link to Month and Day views when there are no events to show.

### PR DESCRIPTION
**Pull request by Camwyn**
Was honestly unsure if this should get lumped into the past events bucket or just merged into the sprint on it's own? On hold until I am sure this is where it needs to go.

Related: https://github.com/moderntribe/events-pro/pull/1533

📷 (month) http://p.tri.be/6Pdoyg
📷 (day) http://p.tri.be/Ht7fbh

[TEC-3400]

[TEC-3400]: https://moderntribe.atlassian.net/browse/TEC-3400

### CODE REVIEW

File: src/Tribe/Views/V2/Messages.php [hunk 1]
```diff
@@ -90,7 +90,14 @@ public function __construct( $render_strategy = null, array $messages = [] ) {
 	 */
 	public static function for_key( $key, ...$values ) {
 		$map = [
-			'no_results_found'                 => __( 'There were no results found.', 'the-events-calendar' ),
+			'no_results_found'                 => __(
+				'There were no results found.',
+				'the-events-calendar'
+			),
+			'month_no_results_found'           => __(
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Messages.php:97]: :no_entry_sign: A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders.

<sub>hash: 35afb13866113225a595c537bda9a249</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 2]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+	 *
+	 * @param bool  $canonical         Whether to return the canonical (pretty) version of the URL or not.
+	 * @param array $passthru_vars     An optional array of query variables that should pass thru the method untouched
+	 *                                 in key and value.
+	 * @return string                  The html link and message.
+	 */
+	public function fast_forward( $canonical = false, array $passthru_vars = [] ) {
+		if ( ! empty( $this->context->get( 'tribe-bar-date' ) ) ) {
+			$date = $this->context->get( 'tribe-bar-date' );
+		} else if ( ! empty( $this->context->get( 'event_date' ) ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:229]: :warning: Usage of ELSE IF is discouraged; use ELSEIF instead

<sub>hash: 3d5792e0fa71bbc1cda133d16192d94b</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 3]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+			$date = strtok( $this->context->get( 'today' ), ' ' );
+		}
+
+		$cache_key = __METHOD__ . '_' . md5( wp_json_encode( array_merge( [ $date ] ) ) );
+
+		if ( isset( $this->cached_urls[ $cache_key ] ) ) {
+			return $this->cached_urls[ $cache_key ];
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page(1)->get_ids();
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:241]: * :no_entry_sign: Expected 1 spaces after opening bracket; 0 found
* :no_entry_sign: Expected 1 spaces before closing bracket; 0 found
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis

<sub>hash: 31b1c69fc5ea6c1f82b34108c7747b47</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 4]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+		$url_date   = Dates::build_date_object( $next_event->start_date );
+		$url        = $this->build_url_for_date( $url_date, $canonical, $passthru_vars );
+
+		$this->cached_urls[ $cache_key ] = $url;
+
+
+		$link = sprintf(
+			/* translators: 1: Name of a city 2: ZIP code */
+			__( 'Jump to the %1$snext upcoming event(s)%2$s.', 'the-events-calendar' ),
+		   '<a href="' . esc_url( $url )  . '">',
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:252]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 12 spaces but found 11
* :no_entry_sign: Concat operator must be surrounded by a single space

<sub>hash: 5e3a595632f88f33b11fc165a6a79725</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 5]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+		   '</a>'
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:253]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 12 spaces but found 11

<sub>hash: f7f22345f1dfacd6568cdcecf441889c</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 6]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+	   );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:254]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 8 spaces but found 7
* :no_entry_sign: Line indented incorrectly; expected at least 2 tabs, found 1

<sub>hash: 03664f83310bbfb6a01dabd8dc6cd204</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 7]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+	 *
+	 * @param bool  $canonical         Whether to return the canonical (pretty) version of the URL or not.
+	 * @param array $passthru_vars     An optional array of query variables that should pass thru the method untouched
+	 *                                 in key and value.
+	 * @return string                  The html link and message.
+	 */
+	public function fast_forward( $canonical = false, array $passthru_vars = [] ) {
+		if ( ! empty( $this->context->get( 'tribe-bar-date' ) ) ) {
+			$date = $this->context->get( 'tribe-bar-date' );
+		} else if ( ! empty( $this->context->get( 'event_date' ) ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:437]: :warning: Usage of ELSE IF is discouraged; use ELSEIF instead

<sub>hash: dc7ca46e502d3272df17a44b9703ed9e</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 8]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+			$date = strtok( $this->context->get( 'today' ), ' ' );
+		}
+
+		$cache_key = __METHOD__ . '_' . md5( wp_json_encode( array_merge( [ $date ] ) ) );
+
+		if ( isset( $this->cached_urls[ $cache_key ] ) ) {
+			return $this->cached_urls[ $cache_key ];
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page(1)->get_ids();
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:449]: * :no_entry_sign: Expected 1 spaces after opening bracket; 0 found
* :no_entry_sign: Expected 1 spaces before closing bracket; 0 found
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis

<sub>hash: 4956c7ea9ab89ec06087b058a088c8f6</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 9]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+		$url_date   = Dates::build_date_object( $next_event->start_date );
+		$url        = $this->build_url_for_date( $url_date, $canonical, $passthru_vars );
+
+		$this->cached_urls[ $cache_key ] = $url;
+
+
+		$link = sprintf(
+			/* translators: 1: Name of a city 2: ZIP code */
+			__( 'Jump to the %1$snext upcoming event(s)%2$s.', 'the-events-calendar' ),
+		   '<a href="' . esc_url( $url )  . '">',
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:460]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 12 spaces but found 11
* :no_entry_sign: Concat operator must be surrounded by a single space

<sub>hash: e7db390c8c2a24e95cd9a4d1e477a3db</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 10]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+		   '</a>'
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:461]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 12 spaces but found 11

<sub>hash: d68ac647016d6f03c45805674792731c</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 11]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+	   );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:462]: * :warning: Found precision alignment of 3 spaces.
* :no_entry_sign: Multi-line function call not indented correctly; expected 8 spaces but found 7
* :no_entry_sign: Line indented incorrectly; expected at least 2 tabs, found 1

<sub>hash: db63838c7c187a63c65bdb15da4103b2</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 12]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+			$date = strtok( $this->context->get( 'today' ), ' ' );
+		}
+
+		$cache_key = __METHOD__ . '_' . md5( wp_json_encode( array_merge( [ $date ] ) ) );
+
+		if ( isset( $this->cached_urls[ $cache_key ] ) ) {
+			return $this->cached_urls[ $cache_key ];
+		}
+
+		$next_event = ( array ) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->get_ids();
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:241]: :no_entry_sign: Cast statements must not contain whitespace; expected "(array)" but found "( array )"

<sub>hash: bc416b54257a0950d99f3f22dcce0cd4</sub>

File: src/Tribe/Template/Day.php [hunk 13]
```diff
@@ -152,6 +152,7 @@ protected function nothing_found_notice() {
 				Tribe__Notices::set_notice( 'events-not-found', sprintf( esc_html__( 'No matching %1$s listed under %2$s scheduled for %3$s. Please try another day.', 'the-events-calendar' ), $events_label_plural_lowercase, $tax_term, '<strong>' . date_i18n( tribe_get_date_format( true ), strtotime( get_query_var( 'eventDate' ) ) ) . '</strong>' ) );
 			} elseif ( empty( $search_term ) && empty( $geographic_term ) ) {
 				Tribe__Notices::set_notice( 'events-not-found', sprintf( esc_html__( 'No %1$s scheduled for %2$s. Please try another day.', 'the-events-calendar' ), $events_label_plural_lowercase, '<strong>' . date_i18n( tribe_get_date_format( true ), strtotime( get_query_var( 'eventDate' ) ) ) . '</strong>' ) );
+
```
Comment (lucatume) [inline src/Tribe/Template/Day.php:155]: Remove this, unless there's a reasoning behind it.

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 14]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
 
+	/**
+	 * Creates a HTML link and "fast forward" message to append to the "no events found" message.
+	 *
+	 * @param bool  $canonical         Whether to return the canonical (pretty) version of the URL or not.
+	 * @param array $passthru_vars     An optional array of query variables that should pass thru the method untouched
+	 *                                 in key and value.
+	 * @return string                  The html link and message.
+	 */
+	public function fast_forward( $canonical = false, array $passthru_vars = [] ) {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Day_View.php:226]: Rename to `get_fast_forward_link`?

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 15]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+		if ( ! empty( $this->context->get( 'tribe-bar-date' ) ) ) {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Day_View.php:227]: The `event_date` Context location is already looking up `tribe-bar-date` before; see [here](https://github.com/moderntribe/the-events-calendar/blob/620e3f295467ad45a9bb7121ced8531c36ba2005/src/Tribe/Service_Providers/Context.php#L87-L97)
Furthermore, the `tribe-bar-date` Context location does not exist.

I would change this whole block to `$date = $this->context->get( 'event_date', $this->context->get( 'today' ) );`

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 16]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+			$date = $this->context->get( 'tribe-bar-date' );
+		} elseif ( ! empty( $this->context->get( 'event_date' ) ) ) {
+			$date = $this->context->get( 'event_date' );
+		} else {
+			$date = strtok( $this->context->get( 'today' ), ' ' );
+		}
+
+		$cache_key = __METHOD__ . '_' . md5( wp_json_encode( array_merge( [ $date ] ) ) );
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Day_View.php:235]: Why the `array_merge` on one element here?

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 17]
```diff
@@ -214,6 +215,47 @@ protected function sort_events( $events ) {
...
+
+		if ( isset( $this->cached_urls[ $cache_key ] ) ) {
+			return $this->cached_urls[ $cache_key ];
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->get_ids();
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Day_View.php:241]:
```suggestion
		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->fields( 'ids' )->first();
```

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 18]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
 
+	/**
+	 * Creates a HTML link and "fast forward" message to append to the "no events found" message.
+	 *
+	 * @param bool  $canonical         Whether to return the canonical (pretty) version of the URL or not.
+	 * @param array $passthru_vars     An optional array of query variables that should pass thru the method untouched
+	 *                                 in key and value.
+	 * @return string                  The html link and message.
+	 */
+	public function fast_forward( $canonical = false, array $passthru_vars = [] ) {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Month_View.php:434]: Rename this to `get_fast_forward_link`?

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 19]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+		if ( ! empty( $this->context->get( 'tribe-bar-date' ) ) ) {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Month_View.php:435]: Apply here the same refactoring logic as the other one.

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 20]
```diff
@@ -423,6 +423,47 @@ protected function get_url_date_format() {
...
+			$date = strtok( $this->context->get( 'today' ), ' ' );
+		}
+
+		$cache_key = __METHOD__ . '_' . md5( wp_json_encode( array_merge( [ $date ] ) ) );
+
+		if ( isset( $this->cached_urls[ $cache_key ] ) ) {
+			return $this->cached_urls[ $cache_key ];
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->get_ids();
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Month_View.php:449]: Same `first` refactoring here.

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 21]
```diff
@@ -214,6 +214,46 @@ protected function sort_events( $events ) {
...
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->fields( 'ids' )->first();
+		$next_event = tribe_get_event( array_shift( $next_event ) );
+
+		if ( empty( $next_event ) ) {
+			return false;
+		}
+
+		$url_date   = Dates::build_date_object( $next_event->start_date );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:242]: :warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 3 spaces

<sub>hash: b671f573d104917e3b567a7c9c138370</sub>

File: src/Tribe/Views/V2/Views/Day_View.php [hunk 22]
```diff
@@ -214,6 +214,46 @@ protected function sort_events( $events ) {
...
+		$url        = $this->build_url_for_date( $url_date, $canonical, $passthru_vars );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Day_View.php:243]: :warning: Equals sign not aligned with surrounding assignments; expected 6 spaces but found 8 spaces

<sub>hash: b1089e34160cced801498ef719f9b6d9</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 23]
```diff
@@ -423,6 +423,46 @@ protected function get_url_date_format() {
...
+		}
+
+		$next_event = (array) tribe_events()->where( 'starts_after', $date )->per_page( 1 )->fields( 'ids' )->first();
+		$next_event = tribe_get_event( array_shift( $next_event ) );
+
+		if ( empty( $next_event ) ) {
+			return false;
+		}
+
+		$url_date   = Dates::build_date_object( $next_event->start_date );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:451]: :warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 3 spaces

<sub>hash: db07f34346ec170ae45bf4e354b28d69</sub>

File: src/Tribe/Views/V2/Views/Month_View.php [hunk 24]
```diff
@@ -423,6 +423,46 @@ protected function get_url_date_format() {
...
+		$url        = $this->build_url_for_date( $url_date, $canonical, $passthru_vars );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Month_View.php:452]: :warning: Equals sign not aligned with surrounding assignments; expected 6 spaces but found 8 spaces

<sub>hash: d790aa78e69f9cbd6a808defcfa7de19</sub>

File: tests/views_integration/Tribe/Events/Views/V2/Views/__snapshots__/List_ViewTest__test_render_with_upcoming_events__1.php [hunk 25]
```diff
@@ -25,11 +25,15 @@ class="tribe-events-view-loader tribe-common-a11y-hidden"
 	data-js="tribe-events-view-data"
 	type="application/json"
 >
+<<<<<<< HEAD
```
Comment (lucatume) [inline tests/views_integration/Tribe/Events/Views/V2/Views/__snapshots__/List_ViewTest__test_render_with_upcoming_events__1.php:28]: @Camwyn this looks like a merge artifact.

Comment (Camwyn) [reply tests/views_integration/Tribe/Events/Views/V2/Views/__snapshots__/List_ViewTest__test_render_with_upcoming_events__1.php:28]: dangit

**Pull request merged**